man/man1/flagmail.1
.TH FLAGMAIL 1
.\" @(#) $Header: /xtel/pp/pp-beta/man/man1/RCS/flagmail.1,v 6.0 1991/12/18 20:43:41 jpo Rel $
.\"
.\" $Log: flagmail.1,v $
.\" Revision 6.0  1991/12/18  20:43:41  jpo
.\" Release 6.0
.\"
.\"
.\"
.SH NAME
flagmail \- interface to xalert
.SH SYNOPSIS
.in +.5i
.ti -.5i
flagmail
\%[-d] \%[-e\ string] \%[-f\ format] \%[-h\ hostfile] \%[-p num]
\%[-s\ size] \%[-w\ width] \%[\-x]
.in -.5i
.SH DESCRIPTION
Flagmail is an interface to the \fIxalert\fP program, which runs under
the X windowing system. It's task is to build up a alert message from
a message read from the standard input and pass it to the \fIxalert\fP
for display.  In actual fact, \fIflagmail\fP is not limited to talking
to \fIxalert\fP but can talk to any process that understands the
protocol and listens on the appropriate port.
.PP
If it can not contact such an X server, then it will attempt to locate
the users terminal line and notify that in the same manner as
\fIttyalert\fP.
.PP
It has several options to control the format.
.TP
\-d
Run in debug mode. This stops the process disappearing into the
background at the first opportunity.
.TP
\-e string
Set the \fIextra\fP variable. This is used in part of the display
string and defaults to \*(lqNEW: \*(rq.
.TP
\-f format
This allows overriding the default format of the message displayed.
.TP
\-h hostfile
This changes the file that is read to discover the host xalert is
running on.
.TP
\-p number
Change the default udp port that is sent to.
.TP
\-s size
This sets the \fIsize\fP variable which can be set in the
\&\fI.mailfilter\fP delivery script (typically using the
\*(lq$(size)\*(rq variable).
.TP
\-w width
This sets the maximum width of the output line. The string is
truncated to this width.
.TP
\-x 
Force the use of protocol.
.PP
The default display string is the following
.sp
.ti +.5i
.nf
%extra(%size) %20from %subject << %50body
.fi
.sp
.SH FILES
.ta \w'~/.alert\0\0'u
~/.alert	The default alert file.
.SH "SEE ALSO"
xalert(1), ttyalert(1)
.br
\fIPP Manual \- Volume 3: User Manual\fP
.SH "BUGS"
The formatting string ought to be more general.


CHANGES









             Changes from PP version 5.2 to 6.0





_1.  _O_v_e_r_v_i_e_w

This  short  document  is  a   description   of   the   more
user/administrator  visible changes that have taken place in
PP between the last interim version (5.2) and the beta  ver-
sion  (6.0). This document is not exhaustive or complete but
is a guide to what to look out for. In general the following
may be assumed.

o+    The Queue files are compatible  between  5.2  and  6.0.
     Note  however that the Queue structure has changed. Old
     channels will have to be recompiled  to  recognise  the
     new format, but new channels will read old queue files.

o+    The qmgr protocol is compatible between 5.2 and 6.0

o+    The syntax of some of  the  table  files  has  changed.
     Where appropriate conversion scripts are provided.

o+    Some procedure calls have been  changed  or  have  dif-
     ferent parameters.

o+    the libdsap.a library is now required for compilation.

In summary, you should be careful when upgrading to the  6.0
release.   It  will  not  work  by  just  replacing  the 5.2
binaries with 6.0. For definite the domain table  will  HAVE
to be changed.

_2.  _D_e_t_a_i_l_e_d _c_h_a_n_g_e_s

This list is in no particular order of preference or  impor-
tance.

Domain Table
     The domain table syntax has been  changed  to  be  more
     explicit  and  has  been  brought into line with the OR
     table syntax. We feel this is clearer and more logical.
     Tools  that  process  the domain table have been taught
     the   new   format   and   there   are    scripts    in
     Tools/tables/compat  to convert between the old and new
     formats.  There is NO backwards compatibility here. Old
     format tables will just plain not work.

Aliases Table



                     December 18, 1991





                           - 2 -


     The aliases table syntax has changed to a  main  choice
     of  synonym or alias with further qualifiers specifying
     the address type of the synonym or alias.  The  changes
     are

          x:alias y         x:alias y
          x:synonym y       x:synonym y
          x:822-synonym y   x:synonym y 822
          x:x400-synonym y  x:synonym y x400
          x:822 y           x:alias y 822
          x:x400 y          x:alias y x400
     where any white space in  y  must  be  quoted  using  a
     backslash  There  is also a "external" qualifier.  This
     disables further lookups of synonyms  and  aliases  and
     can  be  used  for  a  main  site mail hub mapping from
     site_wide_name@localsite                             to
     local_machine_name@localmachine  and  vice  versa.  The
     use of external stops the looping alias  problem  which
     would otherwise occur.

Return of Contents
     Return of contents is now supported.  This  is  a  side
     effect  and  part  of  the  work on routing of delivery
     reports.

Use of pepsy
     Use of the PEPSY ASN.1 compiler rather  than  PEPY/POSY
     is  now  used  in  many  places. This makes for smaller
     binaries and much quicker compilation.

ISODE 7.0
     PP 6.0 is aligned to ISODE 7.0. It will not work  under
     ISODE 6.0.  If you are using the 6.0 release you should
     be using ISODE 7.0. It should however work under  ISODE
     6.8.

Tailor file
     There have been many changes to the  tailor  file.  The
     introduction of several new variables and the splitting
     up of header and body part types. These are

          dap_user     Directory username for DS lookups
          dap_passwd   Directory Passwd for same
          fsync        disable use of fsync(2)
          headertype   list of headers c.f. bodypart
          queuestruct  use hierarchical queue structure
          submit_addr  call submit over network
          diskuse      limit the disk space consumed by PP

     Channel tailoring also has a number of new variables

          drchan       replaces old outchan
          hdrin        headers inbound
          hdrout       headers outbound



                     December 18, 1991





                           - 3 -


          ininfo       inbound channel info
          inadr        inbound channel addr type
          insubadr     inbound channel sub address type
          intable      inbound channel table
          key          give alternative names to channel
          mtatable     list of known mtas for channel binding
          outinfo      outbound channel info
          outsubadr    outbound channel sub address
          outtable     outbound channel table
          solo-proc    channel requires address one at a time
          bad-sender-policyset the policy for unroutable senders
          check        set the checking mode of the channel

Channel pairing
     Channels may now be explicitly  paired  together.  This
     allows one definition to specify both an inbound and an
     outbound channel.  It also allows several logical chan-
     nels  for  one  inbound protocol by selecting a channel
     per MTA.

Timezones
     Many more lettered timezones are recognised. This  list
     is ambiguous at best and numeric forms should really be
     used except at user interfaces.

Delivery Reports
     Delivery reports are now routed back to their  origina-
     tor.  In 5.2 they were returned directly down the chan-
     nel they arrived on. This allows for one way  links  in
     general  makes  things  easier. It also allows protocol
     conversion so that, for instance, a delivery report  on
     an  X.400 message may go back via an RFC-822 if that is
     the only option.  A side effect of this is the  channel
     variable  outchan  has been renamed drchan and has dif-
     ferent semantics attached to it (see the manual).

Header parts
     Separation of headers and bodyparts for  each  channel.
     This  is  mainly a syntactic change but will allow more

Uip/rcvalert/flagmail.c
/* client.c: client for alert program */

# ifndef lint
static char Rcsid[] = "@(#)$Header: /xtel/pp/pp-beta/Uip/rcvalert/RCS/flagmail.c,v 6.0 1991/12/18 20:39:41 jpo Rel $";
# endif

/*
 * $Header: /xtel/pp/pp-beta/Uip/rcvalert/RCS/flagmail.c,v 6.0 1991/12/18 20:39:41 jpo Rel $
 *
 * $Log: flagmail.c,v $
 * Revision 6.0  1991/12/18  20:39:41  jpo
 * Release 6.0
 *
 */



#include <stdio.h>
#include <isode/general.h>
#include <isode/manifest.h>
#include <isode/internet.h>
#include <sys/time.h>
#include <pwd.h>
#include <ctype.h>

#include "data.h"

char	displayline[BUFSIZ];
char	from[BUFSIZ];
char	username[40];
char	hostname[BUFSIZ];
char	homedir[512];
char	*filename = ".alert";
char	*extra = "NEW: ";
char	*format = "%extra(%size) %20from %subject << %80body";
int	width = 80;
int	debug = 0;
int	theport = 0;

char	*myname;

extern char	*index ();

main (argc, argv)
int	argc;
char	*argv[];
{
	extern char	*optarg;
	extern int	optind;
	int	opt;
	int	size = 0;
	int	use_tty = 0;
	int	use_x	= 0;

	myname = argv[0];
	pp_initialise (myname, 0);
	while((opt = getopt(argc, argv, "de:f:h:p:s:txw:")) != EOF)
		switch (opt) {
		    case 'e':
			extra = optarg;
			break;

		    case 'f':
			format = optarg;
			break;

		    case 'd':
			debug = 1;
			break;

		    case 'h':
			filename = optarg;
			break;

		    case 's':
			size = atoi (optarg);
			break;

		    case 'p':
			theport = atoi (optarg);
			break;

		    case 't':
			use_tty = 1;
			break;

		    case 'w':
			width = atoi (optarg);
			break;

		    case 'x':
			use_x = 1;
			break;

		    default:
			fprintf (stderr, "Usage: %s [options]\n",
				 myname);
			break;
		}
	argc -= optind;
	argv += optind;

	init_stuff ();
	
	parse_msg (displayline, from, format, BUFSIZ-1, size, extra);
	chop(displayline);

	if (debug == 0 && fork () > 0)
		exit (1);
	if (use_tty || (notify_x () == 0 && ! use_x))
	{	if (width > 0 && width < sizeof displayline)
			displayline[width] = 0;
		chop(displayline);
		(void) notify_normal (displayline, username);
	}
	exit (1);
}

init_stuff ()
{
	struct passwd *pwd;
	char	*cp, *getenv ();

	if ((pwd = getpwuid (getuid())) == NULL) {
		fprintf (stderr, "No user name\n");
		exit(-1);
	}

	(void) strcpy (username, pwd -> pw_name);
	if ((cp = getenv ("HOME")) != NULL && *cp)
		(void) strcpy (homedir, cp);
	else
		(void) strcpy (homedir, pwd -> pw_dir);
}


notify_x ()
{
	struct data data;
	struct sockaddr_in sin;
	struct hostent *hp;
	int	i;
	int	sd;
	
	if (!findhost(hostname))
		return 0;

	if ((sd = socket (AF_INET, SOCK_DGRAM, 0)) < 0) {
		perror ("socket");
		exit (1);
	}

	bzero ((char *)&sin, sizeof sin);

	sin.sin_family = AF_INET;
	sin.sin_port = htons (theport);

	bzero ((char *)&data, sizeof data);
	data.refid = htonl(getpid ());
	(void) strcpy (data.user, username);
	(void) strncpy (data.data, displayline, sizeof (data.data) - 1);
	data.data[sizeof data.data - 1] = 0;
	(void) strncpy (data.from, from, sizeof(data.from) - 1);
	data.from[sizeof data.from - 1] = 0;

	if ((hp = gethostbyname (hostname)) == NULL) {
		fprintf (stderr, "No such host %s\n", hostname);
		return 0;
	}
	bcopy (hp -> h_addr, (char *)&sin.sin_addr, hp -> h_length);
	for (i = 0; i < N_RETRYS; i++) {
		if (debug) printf("Send to %s.%d\n", hostname, theport);
		if (sendto (sd, (char *)&data, sizeof data, 0,
			    (struct sockaddr *)&sin, sizeof sin) < 0)
			perror ("sendto");

		if (replys (sd))
			return 1;
	}
	if (debug) printf("No reply from %s.%d\n", hostname, theport);
	return 0;
}

replys (sd)
int	sd;
{
	fd_set rfds;
	char	buffer[20];
	struct timeval timer;
	struct sockaddr_in nsin;
	int	len;

	timer.tv_usec = 0;
	timer.tv_sec = 10;
	FD_ZERO (&rfds);
	FD_SET (sd, &rfds);

	if (select (sd + 1, &rfds, NULLFD, NULLFD, &timer) <= 0)
		return 0;

	if (FD_ISSET (sd, &rfds)) {
		len = sizeof nsin;

		if (recvfrom (sd, buffer, 3, 0,
			      (struct sockaddr *)&nsin, &len) < 0)
			perror ("recvfrom");
		return 1;
	}
	if (debug) printf("No reply from %s.%d\n", hostname, theport);
	return 0;
}

findhost (host)
char	*host;
{
	FILE	*fp;
	char	*cp;
	char	buf[BUFSIZ];

	if (*filename == '/' || strncmp (filename, "./", 2) == 0 ||
	    strncmp (filename, "../", 3) == 0)
		(void) strcpy (buf, filename);
	else
		(void) sprintf (buf, "%s/%s", homedir, filename);

	if ((fp = fopen (buf, "r")) == NULL &&
	    (fp = fopen (filename, "r")) == NULL)
		return 0;

	if (fgets (host, BUFSIZ, fp) == NULL) {
		(void) fclose (fp);
		return 0;
	}
	(void) fclose (fp);
	if (cp = index (hostname, '\n'))
		*cp = '\0';
	if (cp = index (hostname, ' ')) {
		*cp++ = '\0';
		if (theport == 0)
			theport = atoi (cp);
	}

	if (theport == 0) {
		fprintf (stderr, "No port!\n");
		return 0;
	}
		
	return 1;
}

chop(string)
char *string;
{	char *end = string + strlen(string) -1;
	while (end != string && (*end == ' ' || *end == '\t')) *end-- = 0;
}
